### PR TITLE
【Paddle Toolkit Development Competition No.2】[Custom OP] Add contiguous api support to paddle::Tensor

### PR DESCRIPTION
**Pull request by BeingGod**
### PR Category

Operator Mechanism

### PR Types

New features

### Description

We find some limitation of paddle custom op. The target of this PR is to improve custom op capability. Specifically as following:
1.  Add contiguous api support to paddle::Tensor

### CODE REVIEW

**Comment by BeingGod** [comment]
PTAL. Thx ! @HydrogenSulfate

File: test/custom_op/test_custom_contiguous.py [hunk 1]
```diff
@@ -0,0 +1,73 @@
...
+import os
+import unittest
+
+import numpy as np
+from utils import (
+    extra_cc_args,
+    extra_nvcc_args,
+    paddle_includes,
+    paddle_libraries,
+)
```
Comment (HydrogenSulfate) [inline test/custom_op/test_custom_contiguous.py:24]: 这行代码在DCU上会报错，建议直接复制<https://github.com/PaddlePaddle/Paddle/blob/develop/test/custom_op/utils.py>和<https://github.com/PaddlePaddle/Paddle/blob/develop/python/paddle/utils/cpp_extension/extension_utils.py>里的相关代码过来，手动构造这四个变量，避免from utils的导入行为引起的报错

Comment (BeingGod) [reply test/custom_op/test_custom_contiguous.py:24]: Ack. Thx.

File: test/custom_op/test_custom_cast_op_jit.py [hunk 2]
```diff
@@ -57,6 +57,9 @@ def custom_cast_dynamic(device, dtype, np_x):
 
     out.backward()
 
+    # NOTE: debug code to trigger UT
+    print("xxxxx")
```
Comment (HydrogenSulfate) [inline test/custom_op/test_custom_cast_op_jit.py]: 此处记得删掉，否则会被CI检测到而挂掉

**Review by HydrogenSulfate — COMMENTED** [review]
覆盖率和几个CI没过，可以再看下

File: test/custom_op/test_custom_contiguous.py (hunk 1 above)
Comment (HydrogenSulfate) [inline test/custom_op/test_custom_contiguous.py:24]: 这里记得改一下

**Comment by BeingGod** [comment]
@HydrogenSulfate This PR is ready. Can you help me merge it ?

File: paddle/phi/api/include/tensor.h [hunk 3]
```diff
@@ -630,6 +630,24 @@ class PADDLE_API Tensor final {
...
+   * @brief Determine whether tensor is contiguous
+   *
+   * @return bool
+   */
+  bool is_contiguous() const;
+
+  /**
+   * @brief Returns a contiguous in memory tensor containing the same data as
+   * current Tensor. If self tensor is already contiguous, this function returns
+   * the current Tensor.
```
Comment (zyfncg) [inline paddle/phi/api/include/tensor.h:652]: 看下面的实现里return的都是当前Tensor, 差别在于是否更新内部的DenseTensor，这更像是一个inplace接口，inplace接口容易有潜在风险，是否可以直接返回一个新Tensor?

Comment (BeingGod) [reply paddle/phi/api/include/tensor.h:652]: Ack. Thx.

**Comment by BeingGod** [comment]
@zyfncg This PR is ready. Can you help me merge it ?

**Comment by BeingGod** [comment]
To custom operators the input tensors are contiguous so we don't need it temporary.
image

**Comment by BeingGod** [comment]
> To custom operators the input tensors are contiguous so we don't need it temporary. image

In custom operator user may call `paddle::experimental::transpose` (stride kernel) it will cause tensor becomes non-contiguous. 
So it's necessary to implement `is_contiguous` and `contiguous` function to `paddle::Tensor` class.

**Review by zyfncg — APPROVED** [review]

**Review by risemeup1 — APPROVED** [review]

**Review by XieYunshen — APPROVED** [review]

**Pull request merged**